ThreeBotPackages/tfgrid_solutions/tfgrid_solutions/chatflows/four_to_six_gateway.py

```diff
@@ -1,79 +1,101 @@
-
-    resv_id = j.sal.reservation_chatflow.reservation_register_and_pay(
-        reservation, expiration, customer_tid=j.me.tid, currency=currency, bot=bot
-    )
-    reservation_result = j.sal.reservation_chatflow.reservation_wait(bot, resv_id)
-
-    j.sal.reservation_chatflow.reservation_save(
-        resv_id, f"4to6GW:{resv_id}", "tfgrid.solutions.4to6gw.1", user_form_data
-    )
-
-    res = """
-            # Use the following template to configure your wireguard connection. This will give you access to your network.
-            ## Make sure you have <a target="_blank" href="https://www.wireguard.com/install/">wireguard</a> installed
-            Click next
-            to download your configuration
-            """
-    bot.md_show(j.core.text.strip(res))
-
-    cfg = reservation_result[0].data_json
-    wgconfigtemplate = """\
-    [Interface]
-    Address = {{cfg.ips[0]}}
-    PrivateKey = {{privatekey}}
-    {% for peer in cfg.peers %}
-    [Peer]
-    PublicKey = {{peer.public_key}}
-    AllowedIPs = {{",".join(peer.allowed_ips)}}
-    {% if peer.endpoint -%}
-    Endpoint = {{peer.endpoint}}
-    {% endif %}
-    {% endfor %}
-    """
-    config = j.tools.jinja2.template_render(text=wgconfigtemplate, cfg=cfg, privatekey=privatekey.decode())
-    config = j.core.text.strip(config)
-
-    filename = "wg-{}.conf".format(resv_id)
-    bot.download_file(config, filename)
-    res = """
-            # In order to connect to the 4 to 6 gateway execute this command:
-            ## ```wg-quick up ./{}```
-            """.format(
-        filename
-    )
-
-    bot.md_show(j.core.text.strip(res))
+
+    @j.baseclasses.chatflow_step(title="Gateway")
+    def gateway_start(self):
+        user_info = self.user_info()
+        j.sal.reservation_chatflow.validate_user(user_info)
+
+        self.gateway = j.sal.reservation_chatflow.gateway_select(self)
+        self.gateway_id = self.gateway.node_id
+        self.user_form_data["Gateway"] = self.gateway_id
+
+    @j.baseclasses.chatflow_step(title="Gateway expiration time")
+    def expiration_time(self):
+        self.expiration = self.datetime_picker(
+            "Please enter solution expiration time.",
+            required=True,
+            min_time=[3600, "Date/time should be at least 1 hour from now"],
+            default=j.data.time.epoch + 3900,
+        )
+        self.user_form_data["Expiration"] = j.data.time.secondsToHRDelta(self.expiration - j.data.time.epoch)
+
+    @j.baseclasses.chatflow_step(title="Wireguard public key")
+    def wireguard_public_get(self):
+        self.publickey = self.string_ask(
+            "Please enter wireguard public key or leave empty if you want us to generate one for you."
+        )
+        self.privatekey = "enter private key here"
+
+    @j.baseclasses.chatflow_step(title="Create your Wireguard ", disable_previous=True)
+    def wg_reservation(self):
```

Comment:
If we can't go back from this point the previous step should probably mention something that we are going to deploy his stuff now...